Fix indentation in WebTools/Root

amaltaro opened the pull request:
Fixes #10207 

#### Status
not-tested

#### Description
Fix indentation and also make that rbytes assignment more clear, to me at least.

#### Is it backward compatible (if not, which system it affects?)
yes

#### Related PRs
none

#### External dependencies / deployment changes
none

amaltaro committed:
Fix indentation in WebTools/Root

explicit assignment

vkuznet commented on src/python/WMCore/WebTools/Root.py:
```diff
@@ -93,16 +93,17 @@ def access(self):
-                        and request.rfile.rfile.bytes_read) or "-"
+                # cherrypy patch not always available, hence the test
+                if request.rfile.rfile.bytes_read:
+                    rbytes = request.rfile.rfile.bytes_read
```
Alan, I don't think it is safe to assume that request may have rfile, and rfile may have bytes_read data. I think getattr is better approach. But what you probably need is the following:
```
rfile = None
if hasattr(request, "rfile"):
   rfile = request.rfile
  if hasattr(rfile, "rfile"):
     rfile = rfile.rfile
bytes_read = None
if rfile and hasattr(rfile, "bytes_read"):
   bytes_read = getattr(rfile, "bytes_read")
if bytes_read:
   rbytes = bytes_read
else:
   rbytes = "-"
```

I suggest to provide proper unit test to pass different structure of request, e.g. if it is real cherrypy request, if it is None, etc.

amaltaro replied:
That nested getattr only make things unclear. Given that we are inside a try/except block, we can just try to access that attribute directly, which will fall into the except block if it does not exist.

vkuznet commented on src/python/WMCore/WebTools/Root.py:
```diff
@@ -93,16 +93,17 @@ def access(self):
-                        and getattr(request.rfile.rfile, "bytes_read", None)
-                        and request.rfile.rfile.bytes_read) or "-"
+                # cherrypy patch not always available, hence the test
+                if request.rfile.rfile.bytes_read:
```
This is not safe as it will throw exception if any nested member data does not exist. You cannot use if for that. Instead, you should walk through every structure and identify if there is specific data you request in it, see my suggestion below.

vkuznet replied:
It is fine with me, I just pointed out to potential problem. Anyway, I still suggest to add proper unit test to test all possible scenarios how to access rbytes value and check that no identation error will be thrown.

amaltaro committed:
fix unit tests

amaltaro commented:
Thanks for your comments, @vkuznet . It looks like the WebTools/Root unit tests were broken (missing an import). I fixed that and added another unit test and everything passes in my container.

The interesting part is, if I fix the unit tests and ran them against master, all of them succeed and there are no exceptions (neither indentation errors), so I'm not totally sure whether this is going to fix the problem reported...

@yuyiguo Yuyi, any chance you could try to apply this patch to one of the cmsweb-testbed (or dev) pods and see whether it clears those errors?

amaltaro commented:
test this please

amaltaro committed:
first exit, then stop cherrypy thread

commenting out the cherrypy exit bit

fix test docstrings

amaltaro commented:
I haven't yet managed to find a way around those 2 microservice unit tests failing now, even though nothing was changed on them...

@yuyiguo Yuyi, I don't plan to make any other changes to this source code, so feel free to test it and let me know whether it solves or not the problem.

amaltaro commented:
test this please

amaltaro commented:
test this please

amaltaro committed:
add some print debugging

amaltaro committed:
randomize port to listen to

amaltaro committed:
change MS test to port 9100 + randon integer

amaltaro committed:
use standard port 8080; disable autoreload

amaltaro commented:
test this please

amaltaro committed:
exit/stop cherrypy inside each test

amaltaro commented:
test this please

amaltaro committed:
mark tests as integration

re-enable testGetStatus test

amaltaro committed:
merge info and status tests together

amaltaro commented:
I just can't get these unit tests to work, no matter what I do... I might simply mark them as integration and move on, 12 new unit tests passing, 2 "deleted" ...

amaltaro commented:
test this please

amaltaro commented:
test this please

amaltaro committed:
rebase after a long time

amaltaro commented:
I did not manage to get the unit test working with these changes, and given that it is stale since 2021/2022, I won't bother with this PR anymore. Closing this out
